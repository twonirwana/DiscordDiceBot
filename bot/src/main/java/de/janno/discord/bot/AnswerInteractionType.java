package de.janno.discord.bot;

public enum AnswerInteractionType {
    none,
    hidden,
    reroll
}
