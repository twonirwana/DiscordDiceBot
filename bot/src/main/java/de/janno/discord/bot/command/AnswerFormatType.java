package de.janno.discord.bot.command;

public enum AnswerFormatType {
    full,
    without_expression,
    only_dice,
    compact,
    minimal
}
