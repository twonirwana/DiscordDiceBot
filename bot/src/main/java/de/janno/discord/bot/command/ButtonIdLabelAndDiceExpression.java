package de.janno.discord.bot.command;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.NonNull;
import lombok.Value;

@Value
public class ButtonIdLabelAndDiceExpression {
    @NonNull
    String buttonId;
    @NonNull
    String label;
    @NonNull
    String diceExpression;

    @JsonCreator
    public ButtonIdLabelAndDiceExpression(
            @JsonProperty("buttonId") @NonNull String buttonId,
            @JsonProperty("label") @NonNull String label,
            @JsonProperty("diceExpression") @NonNull String diceExpression) {
        this.buttonId = buttonId;
        this.label = label;
        this.diceExpression = diceExpression;
    }

    public String toShortString() {
        if (diceExpression.equals(label)) {
            return diceExpression;
        }
        return String.format("%s@%s", diceExpression, label);
    }
}

