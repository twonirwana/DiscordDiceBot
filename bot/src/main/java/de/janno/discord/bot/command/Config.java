package de.janno.discord.bot.command;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import lombok.ToString;

import java.io.Serializable;

/**
 * A configuration for a dice system. It will be created with the slash command and not modified afterwards.
 */
@EqualsAndHashCode
@Getter
@ToString
public class Config implements Serializable {

    private final Long answerTargetChannelId;

    @NonNull
    private final AnswerFormatType answerFormatType;

    @JsonCreator
    public Config(@JsonProperty("answerTargetChannelId") Long answerTargetChannelId,
                  @JsonProperty("answerFormatType") AnswerFormatType answerFormatType) {
        this.answerTargetChannelId = answerTargetChannelId;
        this.answerFormatType = answerFormatType == null ? AnswerFormatType.full : answerFormatType;
    }

    public String toShortString() {
        return String.format("[%s, %s]", getTargetChannelShortString(), answerFormatType);
    }

    protected String getTargetChannelShortString() {
        return answerTargetChannelId == null ? "local" : "target";
    }

}
