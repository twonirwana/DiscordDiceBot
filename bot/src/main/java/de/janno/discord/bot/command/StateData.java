package de.janno.discord.bot.command;

import lombok.EqualsAndHashCode;

import java.io.Serializable;


@EqualsAndHashCode
public class StateData implements Serializable {

    private static final StateData EMPTY = StateData.empty();

    protected StateData() {
    }

    public static StateData empty() {
        return EMPTY;
    }

    public String getShortStringValues() {
        return "";
    }

    public String toShortString() {
        return "[]";
    }
}
