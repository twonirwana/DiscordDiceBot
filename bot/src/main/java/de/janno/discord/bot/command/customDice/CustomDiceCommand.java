package de.janno.discord.bot.command.customDice;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import de.janno.discord.bot.BotMetrics;
import de.janno.discord.bot.command.*;
import de.janno.discord.bot.dice.Dice;
import de.janno.discord.bot.dice.DiceParser;
import de.janno.discord.bot.dice.DiceParserSystem;
import de.janno.discord.bot.dice.DiceSystemAdapter;
import de.janno.discord.bot.persistance.Mapper;
import de.janno.discord.bot.persistance.MessageDataDAO;
import de.janno.discord.bot.persistance.MessageDataDTO;
import de.janno.discord.connector.api.BottomCustomIdUtils;
import de.janno.discord.connector.api.ButtonEventAdaptor;
import de.janno.discord.connector.api.message.ButtonDefinition;
import de.janno.discord.connector.api.message.ComponentRowDefinition;
import de.janno.discord.connector.api.message.EmbedDefinition;
import de.janno.discord.connector.api.message.MessageDefinition;
import de.janno.discord.connector.api.slash.CommandDefinitionOption;
import de.janno.discord.connector.api.slash.CommandInteractionOption;
import de.janno.evaluator.dice.random.NumberSupplier;
import de.janno.evaluator.dice.random.RandomNumberSupplier;
import lombok.NonNull;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

@Slf4j
public class CustomDiceCommand extends AbstractCommand<CustomDiceConfig, StateData> {

    private static final String COMMAND_NAME = "custom_dice";
    private static final List<String> LEGACY_DICE_COMMAND_OPTIONS_IDS = IntStream.range(1, 25).mapToObj(i -> i + "_button").toList();
    private static final String LEGACY_START_ACTION = "legacy_start";
    private static final String BUTTONS_COMMAND_OPTIONS_ID = "buttons";
    private static final String LABEL_DELIMITER = "@";
    private static final String BUTTON_MESSAGE = "Click on a button to roll the dice";
    private static final String CONFIG_TYPE_ID = "CustomDiceConfig";
    private final DiceSystemAdapter diceSystemAdapter;

    public CustomDiceCommand(MessageDataDAO messageDataDAO) {
        this(messageDataDAO, new DiceParser(), new RandomNumberSupplier(), 1000);
    }

    @VisibleForTesting
    public CustomDiceCommand(MessageDataDAO messageDataDAO, Dice dice, NumberSupplier numberSupplier, int maxRolls) {
        super(messageDataDAO);
        this.diceSystemAdapter = new DiceSystemAdapter(numberSupplier, maxRolls, dice);
    }

    @Override
    protected Optional<ConfigAndState<CustomDiceConfig, StateData>> getMessageDataAndUpdateWithButtonValue(long channelId,
                                                                                                           long messageId,
                                                                                                           @NonNull String buttonValue,
                                                                                                           @NonNull String invokingUserName) {
        final Optional<MessageDataDTO> messageDataDTO = messageDataDAO.getDataForMessage(channelId, messageId);
        if (messageDataDTO.isEmpty()) {
            return Optional.empty();
        }
        return Optional.of(deserializeAndUpdateState(messageDataDTO.get(), buttonValue));
    }

    @VisibleForTesting
    ConfigAndState<CustomDiceConfig, StateData> deserializeAndUpdateState(@NonNull MessageDataDTO messageDataDTO, @NonNull String buttonValue) {
        Preconditions.checkArgument(CONFIG_TYPE_ID.equals(messageDataDTO.getConfigClassId()), "Unknown configClassId: %s", messageDataDTO.getConfigClassId());
        return new ConfigAndState<>(messageDataDTO.getConfigUUID(),
                Mapper.deserializeObject(messageDataDTO.getConfig(), CustomDiceConfig.class),
                new State<>(buttonValue, StateData.empty()));
    }

    @Override
    public Optional<MessageDataDTO> createMessageDataForNewMessage(@NonNull UUID configUUID,
                                                                   long guildId,
                                                                   long channelId,
                                                                   long messageId,
                                                                   @NonNull CustomDiceConfig config,
                                                                   @Nullable State<StateData> state) {
        return Optional.of(new MessageDataDTO(configUUID, guildId, channelId, messageId, getCommandId(), CONFIG_TYPE_ID,
                Mapper.serializedObject(config),
                Mapper.NO_PERSISTED_STATE, null));
    }

    @Override
    protected @NonNull String getCommandDescription() {
        return "Configure a custom set of dice buttons";
    }

    @Override
    protected @NonNull List<CommandDefinitionOption> getStartOptions() {
        return List.of(CommandDefinitionOption.builder()
                .name(BUTTONS_COMMAND_OPTIONS_ID)
                .description("Define one or more buttons separated by ';'")
                .type(CommandDefinitionOption.Type.STRING)
                .required(true)
                .build());
    }

    @Override
    protected Collection<CommandDefinitionOption> additionalCommandOptions() {
        return List.of(CommandDefinitionOption.builder()
                .name(LEGACY_START_ACTION)
                .description("Old start command")
                .type(CommandDefinitionOption.Type.SUB_COMMAND)
                .options(LEGACY_DICE_COMMAND_OPTIONS_IDS.stream()
                        .map(id -> CommandDefinitionOption.builder()
                                .name(id)
                                .description("xdy for a set of x dice with y sides, e.g. '3d6'")
                                .type(CommandDefinitionOption.Type.STRING)
                                .build())
                        .collect(Collectors.toList()))
                .option(ANSWER_TARGET_CHANNEL_COMMAND_OPTION)
                .build());
    }


    @Override
    protected @NonNull EmbedDefinition getHelpMessage() {
        return EmbedDefinition.builder()
                .description("Creates up to 25 buttons with custom dice expression e.g. '/custom_dice start buttons:3d6;10d10;3d20'. \n" + diceSystemAdapter.getHelpText(DiceParserSystem.DICE_EVALUATOR))
                .build();
    }

    @Override
    protected Set<String> getStartOptionIds() {
        return Set.of(ACTION_START, LEGACY_START_ACTION);
    }

    @Override
    protected @NonNull Optional<String> getStartOptionsValidationMessage(@NonNull CommandInteractionOption options) {
        List<String> diceExpressionWithOptionalLabel = getButtonsFromCommandOption(options).stream()
                .map(ButtonIdAndExpression::getExpression)
                .distinct()
                .collect(Collectors.toList());
        DiceParserSystem diceParserSystem = options.getName().equals(LEGACY_START_ACTION) ? DiceParserSystem.DICEROLL_PARSER : DiceParserSystem.DICE_EVALUATOR;
        return diceSystemAdapter.validateListOfExpressions(diceExpressionWithOptionalLabel, "/custom_dice help", diceParserSystem);

    }

    private List<ButtonIdAndExpression> getButtonsFromCommandOption(@NonNull CommandInteractionOption options) {
        if (LEGACY_START_ACTION.equals(options.getName())) {
            return LEGACY_DICE_COMMAND_OPTIONS_IDS.stream()
                    .flatMap(id -> options.getStringSubOptionWithName(id).stream()
                            .map(e -> new ButtonIdAndExpression(id, e)))
                    .collect(Collectors.toList());
        }
        ImmutableList.Builder<ButtonIdAndExpression> builder = ImmutableList.builder();
        String buttons = options.getStringSubOptionWithName(BUTTONS_COMMAND_OPTIONS_ID).orElseThrow();
        int idCounter = 1;
        for (String button : buttons.split(";")) {
            builder.add(new ButtonIdAndExpression(idCounter++ + "_button", button));
        }
        return builder.build();
    }

    protected @NonNull CustomDiceConfig getConfigFromStartOptions(@NonNull CommandInteractionOption options) {
        if (LEGACY_START_ACTION.equals(options.getName())) {
            BotMetrics.incrementLegacyStartCounter(getCommandId());
            return getConfigOptionStringList(getButtonsFromCommandOption(options), getAnswerTargetChannelIdFromStartCommandOption(options).orElse(null),
                    DiceParserSystem.DICEROLL_PARSER);
        }
        return getConfigOptionStringList(getButtonsFromCommandOption(options), getAnswerTargetChannelIdFromStartCommandOption(options).orElse(null), DiceParserSystem.DICE_EVALUATOR);
    }

    @VisibleForTesting
    CustomDiceConfig getConfigOptionStringList(List<ButtonIdAndExpression> startOptions, Long channelId, DiceParserSystem diceParserSystem) {
        return new CustomDiceConfig(channelId, startOptions.stream()
                .filter(be -> !be.getExpression().contains(BottomCustomIdUtils.CUSTOM_ID_DELIMITER))
                .filter(be -> !be.getExpression().contains(LABEL_DELIMITER) || be.getExpression().split(LABEL_DELIMITER).length == 2)
                .map(be -> {
                    if (be.getExpression().contains(LABEL_DELIMITER)) {
                        String[] split = be.getExpression().split(LABEL_DELIMITER);
                        return new ButtonIdLabelAndDiceExpression(be.getButtonId(), split[1].trim(), split[0].trim());
                    }
                    return new ButtonIdLabelAndDiceExpression(be.getButtonId(), be.getExpression().trim(), be.getExpression().trim());
                })
                .filter(s -> !s.getDiceExpression().isEmpty())
                .filter(s -> !s.getLabel().isEmpty())
                .filter(lv -> diceSystemAdapter.isValidExpression(lv.getDiceExpression(), diceParserSystem))
                .filter(s -> s.getDiceExpression().length() <= 2000) //limit of the discord message content
                .filter(s -> s.getLabel().length() <= 80) //https://discord.com/developers/docs/interactions/message-components#buttons
                .distinct()
                .limit(25)
                .collect(Collectors.toList()),
                diceParserSystem);
    }

    @Override
    protected @NonNull Optional<EmbedDefinition> getAnswer(CustomDiceConfig config, State<StateData> state) {
        Optional<ButtonIdLabelAndDiceExpression> selectedButton = Optional.ofNullable(state).map(State::getButtonValue)
                .flatMap(bv -> config.getButtonIdLabelAndDiceExpressions().stream()
                        .filter(bld -> bld.getButtonId().equals(bv))
                        .findFirst()
                );
        if (selectedButton.isEmpty()) {
            return Optional.empty();
        }
        //add the label only if it is different from the expression
        final String label = selectedButton.get().getDiceExpression().equals(selectedButton.get().getLabel()) ? null : selectedButton.get().getLabel();
        return Optional.of(diceSystemAdapter.answerRollWithGivenLabel(selectedButton.get().getDiceExpression(), label, false, config.getDiceParserSystem()));
    }

    @Override
    protected @NonNull Optional<MessageDefinition> createNewButtonMessageWithState(CustomDiceConfig config, State<StateData> state) {
        return Optional.of(createNewButtonMessage(config));
    }

    @Override
    public @NonNull MessageDefinition createNewButtonMessage(CustomDiceConfig config) {
        return MessageDefinition.builder()
                .content(BUTTON_MESSAGE)
                .componentRowDefinitions(createButtonLayout(config))
                .build();
    }

    private List<ComponentRowDefinition> createButtonLayout(CustomDiceConfig config) {
        List<ButtonDefinition> buttons = config.getButtonIdLabelAndDiceExpressions().stream()
                .map(d -> ButtonDefinition.builder()
                        .id(BottomCustomIdUtils.createButtonCustomId(getCommandId(), d.getButtonId()))
                        .label(d.getLabel())
                        .build())
                .collect(Collectors.toList());
        return Lists.partition(buttons, 5).stream()
                .map(bl -> ComponentRowDefinition.builder().buttonDefinitions(bl).build())
                .collect(Collectors.toList());
    }

    @Override
    protected @NonNull CustomDiceConfig getConfigFromEvent(@NonNull ButtonEventAdaptor event) {
        String[] split = event.getCustomId().split(BottomCustomIdUtils.LEGACY_CONFIG_SPLIT_DELIMITER_REGEX);
        Long answerTargetChannelId = getOptionalLongFromArray(split, 2);
        Deque<String> buttonIds = new ArrayDeque<>(IntStream.range(1, 26).mapToObj(i -> i + "_button").toList()); //legacy can have 25 buttons

        return new CustomDiceConfig(answerTargetChannelId, event.getAllButtonIds().stream()
                .map(lv -> new ButtonIdLabelAndDiceExpression(buttonIds.pop(), lv.getLabel(), BottomCustomIdUtils.getButtonValueFromLegacyCustomId(lv.getCustomId())))
                .collect(Collectors.toList()), DiceParserSystem.DICEROLL_PARSER);
    }

    @Override
    protected @NonNull State<StateData> getStateFromEvent(@NonNull ButtonEventAdaptor event) {
        final CustomDiceConfig config = getConfigFromEvent(event);
        final String buttonValue = BottomCustomIdUtils.getButtonValueFromLegacyCustomId(event.getCustomId());
        final String mappedButtonValue = config.getButtonIdLabelAndDiceExpressions().stream()
                .filter(bld -> bld.getDiceExpression().equals(buttonValue))
                .map(ButtonIdLabelAndDiceExpression::getButtonId)
                .findFirst()
                .orElseThrow();
        return new State<>(mappedButtonValue, StateData.empty());
    }

    @Override
    public String getCommandId() {
        return COMMAND_NAME;
    }

    @Value
    static class ButtonIdAndExpression {
        @NonNull
        String buttonId;
        @NonNull
        String expression;
    }

}
