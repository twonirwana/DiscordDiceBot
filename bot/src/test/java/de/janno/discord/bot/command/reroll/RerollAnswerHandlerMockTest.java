package de.janno.discord.bot.command.reroll;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.google.common.collect.ImmutableList;
import de.janno.discord.bot.AnswerInteractionType;
import de.janno.discord.bot.ButtonEventAdaptorMock;
import de.janno.discord.bot.ButtonEventAdaptorMockFactory;
import de.janno.discord.bot.command.AnswerFormatType;
import de.janno.discord.bot.command.ButtonIdLabelAndDiceExpression;
import de.janno.discord.bot.command.StateData;
import de.janno.discord.bot.command.IncrementingUUIDSupplier;
import de.janno.discord.bot.command.customDice.CustomDiceCommand;
import de.janno.discord.bot.command.customDice.CustomDiceConfig;
import de.janno.discord.bot.dice.CachingDiceEvaluator;
import de.janno.discord.bot.dice.image.DiceImageStyle;
import de.janno.discord.bot.dice.image.DiceStyleAndColor;
import de.janno.discord.bot.persistance.PersistenceManager;
import de.janno.discord.bot.persistance.PersistenceManagerImpl;
import de.janno.discord.connector.api.message.EmbedOrMessageDefinition;
import de.janno.evaluator.dice.random.RandomNumberSupplier;
import io.avaje.config.Config;
import org.apache.commons.io.FileUtils;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.io.File;
import java.io.IOException;
import java.util.Locale;
import java.util.UUID;
import java.util.function.Supplier;

@ExtendWith(SnapshotExtension.class)
public class RerollAnswerHandlerMockTest {
    PersistenceManager persistenceManager;
    Expect expect;

    @BeforeEach
    void setup() throws IOException {
        File cacheDirectory = new File("imageCache/");
        if (cacheDirectory.exists()) {
            FileUtils.cleanDirectory(cacheDirectory);
        }
        persistenceManager = new PersistenceManagerImpl("jdbc:h2:mem:" + UUID.randomUUID(), null, null);
    }

    @AfterEach
    void cleanUp() throws IOException {
        File cacheDirectory = new File("imageCache/");
        if (cacheDirectory.exists()) {
            FileUtils.cleanDirectory(cacheDirectory);
        }
        Config.setProperty("diceEvaluator.cacheSize", "0");
    }

    @Test
    void customDice_reroll() {
        Supplier<UUID> testingUUIDSupplier = IncrementingUUIDSupplier.create();
        CustomDiceCommand customDiceCommand = new CustomDiceCommand(persistenceManager, new CachingDiceEvaluator(new RandomNumberSupplier(0)), testingUUIDSupplier);
        RerollAnswerHandler rerollAnswerHandler = new RerollAnswerHandler(persistenceManager, new CachingDiceEvaluator(new RandomNumberSupplier(2)), testingUUIDSupplier);
        CustomDiceConfig config = new CustomDiceConfig(null, ImmutableList.of(new ButtonIdLabelAndDiceExpression("1_button", "Dmg", "3d6+5d!4+3d[a/b/c/d]", false, false, null)), AnswerFormatType.full, AnswerInteractionType.reroll, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH, null, null);
        ButtonEventAdaptorMockFactory<CustomDiceConfig, StateData> factory = new ButtonEventAdaptorMockFactory<>("custom_dice", customDiceCommand, config, persistenceManager, false);
        ButtonEventAdaptorMock buttonEvent = factory.getButtonClickOnLastButtonMessage("1_button");

        customDiceCommand.handleComponentInteractEvent(buttonEvent).block();

        expect.scenario("roll").toMatchSnapshot(buttonEvent.getSortedActions());

        EmbedOrMessageDefinition answer = buttonEvent.getSendMessages().getFirst();

        String dieIdButton = answer.getComponentRowDefinitions().getFirst().getComponentDefinitions().getFirst().getId();
        ButtonEventAdaptorMock answerButtonMessage1 = new ButtonEventAdaptorMock(dieIdButton, 2);
        rerollAnswerHandler.handleComponentInteractEvent(answerButtonMessage1).block();
        expect.scenario("select").toMatchSnapshot(answerButtonMessage1.getSortedActions());

        String secondDieIdButton = answer.getComponentRowDefinitions().getFirst().getComponentDefinitions().getLast().getId();
        ButtonEventAdaptorMock answerButtonMessage2 = new ButtonEventAdaptorMock(secondDieIdButton, 2);
        rerollAnswerHandler.handleComponentInteractEvent(answerButtonMessage2).block();
        expect.scenario("secondSelect").toMatchSnapshot(answerButtonMessage2.getSortedActions());

        ButtonEventAdaptorMock answerButtonMessage3 = new ButtonEventAdaptorMock(dieIdButton, 2);
        rerollAnswerHandler.handleComponentInteractEvent(answerButtonMessage3).block();
        expect.scenario("unselect").toMatchSnapshot(answerButtonMessage3.getSortedActions());

        ButtonEventAdaptorMock answerButtonMessage4 = new ButtonEventAdaptorMock(dieIdButton, 2);
        rerollAnswerHandler.handleComponentInteractEvent(answerButtonMessage4).block();
        expect.scenario("reselect").toMatchSnapshot(answerButtonMessage4.getSortedActions());

        String rollIdButton = answer.getComponentRowDefinitions().getLast().getComponentDefinitions().getFirst().getId();
        ButtonEventAdaptorMock answerButtonMessage5 = new ButtonEventAdaptorMock(rollIdButton, 2);
        rerollAnswerHandler.handleComponentInteractEvent(answerButtonMessage5).block();
        expect.scenario("reroll").toMatchSnapshot(answerButtonMessage5.getSortedActions());

        String finishIdButton = answerButtonMessage5.getSendMessages().getFirst().getComponentRowDefinitions().getLast().getComponentDefinitions().getLast().getId();
        ButtonEventAdaptorMock answerButtonMessage6 = new ButtonEventAdaptorMock(finishIdButton, 2);
        rerollAnswerHandler.handleComponentInteractEvent(answerButtonMessage6).block();
        expect.scenario("finish").toMatchSnapshot(answerButtonMessage6.getSortedActions());

    }
}
