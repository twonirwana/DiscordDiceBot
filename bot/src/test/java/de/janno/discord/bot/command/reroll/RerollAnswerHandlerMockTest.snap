de.janno.discord.bot.command.reroll.RerollAnswerHandlerMockTest.customDice_reroll[finish]=[
  [
    "editMessage: message:, buttonValues="
  ]
]


de.janno.discord.bot.command.reroll.RerollAnswerHandlerMockTest.customDice_reroll[reroll]=[
  [
    "editMessage: message:, buttonValues=",
    "sendMessage: EmbedOrMessageDefinition(title=1: Dmg ⇒ 1, 4, 4, 1, 2, 4, 1, 4, 1, 2, d, b, d, descriptionOrContent=3d6+5d!4+3d[a/b/c/d]: [1, 4, 4] [1, 2, 4, 1, 4, 1, 2] [d, b, d], fields=[], componentRowDefinitions=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=1 ∈ d6, id=reroll_answer\u001E1de0i0r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=4 ∈ d6, id=reroll_answer\u001E1de0i1r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=4 ∈ d6, id=reroll_answer\u001E1de0i2r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=1 ∈ d4, id=reroll_answer\u001E5d!e0i0r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=2 ∈ d4, id=reroll_answer\u001E5d!e0i1r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=4 ∈ d4, id=reroll_answer\u001E5d!e0i2r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=1 ∈ d4, id=reroll_answer\u001E5d!e0i2r1\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=4 ∈ d4, id=reroll_answer\u001E5d!e0i3r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=1 ∈ d4, id=reroll_answer\u001E5d!e0i3r1\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=2 ∈ d4, id=reroll_answer\u001E5d!e0i4r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=d ∈ d[a/b/c/d], id=reroll_answer\u001E10de0i0r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=b ∈ d[a/b/c/d], id=reroll_answer\u001E10de0i1r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false), ButtonDefinition(label=d ∈ d[a/b/c/d], id=reroll_answer\u001E10de0i2r0\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SECONDARY, disabled=false)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Roll, id=reroll_answer\u001Eroll\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=SUCCESS, disabled=true), ButtonDefinition(label=Finish, id=reroll_answer\u001Efinish\u001E5bf2071a-bc5d-37ff-8240-7e5ab439d655, style=DANGER, disabled=false)])], hasImage=false, type=EMBED, userReference=true, sendToOtherChannelId=null)"
  ]
]


de.janno.discord.bot.command.reroll.RerollAnswerHandlerMockTest.customDice_reroll[reselect]=[
  [
    "editMessage: message:, buttonValues=1de0i0r0,1de0i1r0,1de0i2r0,5d!e0i0r0,5d!e0i1r0,5d!e0i2r0,5d!e0i2r1,5d!e0i3r0,5d!e0i3r1,5d!e0i4r0,10de0i0r0,10de0i1r0,10de0i2r0,roll,finish"
  ]
]


de.janno.discord.bot.command.reroll.RerollAnswerHandlerMockTest.customDice_reroll[roll]=[
  [
    "deleteMessageById: 0",
    "editMessage: message:processing ..., buttonValues=",
    "sendMessage: EmbedOrMessageDefinition(title=Dmg ⇒ 2, 4, 4, 1, 3, 4, 1, 4, 1, 2, d, b, d, descriptionOrContent=3d6+5d!4+3d[a/b/c/d]: [2, 4, 4] [1, 3, 4, 1, 4, 1, 2] [d, b, d], fields=[], componentRowDefinitions=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=2 ∈ d6, id=reroll_answer\u001E1de0i0r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=4 ∈ d6, id=reroll_answer\u001E1de0i1r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=4 ∈ d6, id=reroll_answer\u001E1de0i2r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=1 ∈ d4, id=reroll_answer\u001E5d!e0i0r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=3 ∈ d4, id=reroll_answer\u001E5d!e0i1r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=4 ∈ d4, id=reroll_answer\u001E5d!e0i2r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=1 ∈ d4, id=reroll_answer\u001E5d!e0i2r1\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=4 ∈ d4, id=reroll_answer\u001E5d!e0i3r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=1 ∈ d4, id=reroll_answer\u001E5d!e0i3r1\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=2 ∈ d4, id=reroll_answer\u001E5d!e0i4r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=d ∈ d[a/b/c/d], id=reroll_answer\u001E10de0i0r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=b ∈ d[a/b/c/d], id=reroll_answer\u001E10de0i1r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false), ButtonDefinition(label=d ∈ d[a/b/c/d], id=reroll_answer\u001E10de0i2r0\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SECONDARY, disabled=false)]), ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Roll, id=reroll_answer\u001Eroll\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=SUCCESS, disabled=true), ButtonDefinition(label=Finish, id=reroll_answer\u001Efinish\u001E4ae71336-e44b-39bf-b9d2-752e234818a5, style=DANGER, disabled=false)])], hasImage=false, type=EMBED, userReference=true, sendToOtherChannelId=null)",
    "sendMessage: EmbedOrMessageDefinition(title=null, descriptionOrContent=Click on a button to roll the dice, fields=[], componentRowDefinitions=[ComponentRowDefinition(buttonDefinitions=[ButtonDefinition(label=Dmg, id=custom_dice\u001E1_button\u001E00000000-0000-0000-0000-000000000000, style=PRIMARY, disabled=false)])], hasImage=false, type=MESSAGE, userReference=false, sendToOtherChannelId=null)"
  ]
]


de.janno.discord.bot.command.reroll.RerollAnswerHandlerMockTest.customDice_reroll[secondSelect]=[
  [
    "editMessage: message:, buttonValues=1de0i0r0,1de0i1r0,1de0i2r0,5d!e0i0r0,5d!e0i1r0,5d!e0i2r0,5d!e0i2r1,5d!e0i3r0,5d!e0i3r1,5d!e0i4r0,10de0i0r0,10de0i1r0,10de0i2r0,roll,finish"
  ]
]


de.janno.discord.bot.command.reroll.RerollAnswerHandlerMockTest.customDice_reroll[select]=[
  [
    "editMessage: message:, buttonValues=1de0i0r0,1de0i1r0,1de0i2r0,5d!e0i0r0,5d!e0i1r0,5d!e0i2r0,5d!e0i2r1,5d!e0i3r0,5d!e0i3r1,5d!e0i4r0,10de0i0r0,10de0i1r0,10de0i2r0,roll,finish"
  ]
]


de.janno.discord.bot.command.reroll.RerollAnswerHandlerMockTest.customDice_reroll[unselect]=[
  [
    "editMessage: message:, buttonValues=1de0i0r0,1de0i1r0,1de0i2r0,5d!e0i0r0,5d!e0i1r0,5d!e0i2r0,5d!e0i2r1,5d!e0i3r0,5d!e0i3r1,5d!e0i4r0,10de0i0r0,10de0i1r0,10de0i2r0,roll,finish"
  ]
]